### PR TITLE
Refactor/remember intersection shapes

### PR DESCRIPTION
**Pull request by MoKob**
# Issue

https://github.com/Project-OSRM/osrm-backend/issues/3313

During the PR, I've kept the old implementation around to make sure the results remain the same. I've found a few problematic scenarios along the way, which required adjustments to the coordinate extraction (border cases, currently handled incorrectly).

As a side-effect, this PR also fixes https://github.com/Project-OSRM/osrm-backend/issues/3241

## Tasklist
 - [x] find source of potential memory corruption after rebase
 - [x] requires https://github.com/Project-OSRM/osrm-backend/pull/3315 merged
 - [x] review
 - [x] adjust for comments

## Requirements / Relations
 based on https://github.com/Project-OSRM/osrm-backend/pull/3315

### CODE REVIEW

**Comment by MoKob** [comment]
Impact of this PR: Measured on the German Road Network

Master:
[info] Renumbering edges: 0.267233s
[info] Generating nodes: 14.4786s
[info] Generating edges: 436.623s

Remembering Shapes:
[info] Renumbering edges: 0.261475s
[info] Generating nodes: 14.634s
[info] Generating edges: 310.801s

**Comment by MoKob** [comment]
On Germany, I currently see a failure writing [here](https://github.com/Project-OSRM/osrm-backend/pull/3341/files#diff-e9ebe6db6549f062de82360a303fe3faR367).

The details are in [this gist](https://gist.github.com/MoKob/405e4ab6f5a81fcb42499a1a5e0a203f). It looks like the iterator might have gotten invalidated, but I can't see a reason for it yet.

**Comment by MoKob** [comment]
The problem in combining both PRs for refactoring on performance is that the benefits from https://github.com/Project-OSRM/osrm-backend/pull/3315 are mostly hidden due to the improved processing order. This sadly results in only a minor benefit from adding the additional refactoring from https://github.com/Project-OSRM/osrm-backend/pull/3315 into this PR. When only running these changes, we see:

[info] Timing statistics for edge-expanded graph:
[info] Renumbering edges: 0.250557s
[info] Generating nodes: 14.6485s
[info] Generating edges: 340.08s.

The additional changes from #3315 only give an additional 15 seconds :(

File: src/engine/guidance/post_processing.cpp [hunk 1]
```diff
@@ -92,37 +92,6 @@ bool compatible(const RouteStep &lhs, const RouteStep &rhs) { return lhs.mode ==
 // invalidate a step and set its content to nothing
 void invalidateStep(RouteStep &step) { step = getInvalidRouteStep(); }
 
-// Compute the angle between two bearings on a normal turn circle
```
Comment (MoKob) [inline src/engine/guidance/post_processing.cpp]: moved to a general utility function

**Comment by MoKob** [comment]
Managed to squeeze out another 40 seconds by using the `simple_mode` for degree <= 2 (traffic lights...)

```
[info] Generating nodes: 14.6197s
[info] Generating edges: 301.669s
```

File: features/guidance/collapse.feature [hunk 2]
```diff
@@ -353,13 +353,13 @@ Feature: Collapse
...
-            | d,g       | first,second,second | depart,turn right,arrive       |
-            | g,f       | second,first,first  | depart,turn right,arrive       |
-            | g,c       | second,first,first  | depart,end of road left,arrive |
+            | waypoints | route               | turns                        |
+            | d,c       | first,first,first   | depart,continue uturn,arrive |
+            | a,f       | first,first,first   | depart,continue uturn,arrive |
+            | a,g       | first,second,second | depart,turn left,arrive      |
+            | d,g       | first,second,second | depart,turn right,arrive     |
+            | g,f       | second,first,first  | depart,turn right,arrive     |
+            | g,c       | second,first,first  | depart,turn left,arrive      |
```
Comment (MoKob) [inline features/guidance/collapse.feature]: due to https://github.com/Project-OSRM/osrm-backend/issues/3365

File: include/extractor/extraction_way.hpp [hunk 3]
```diff
@@ -15,19 +15,18 @@ namespace extractor
...
-    inline void maybeSetString(std::string &str, const char* value)
+inline void maybeSetString(std::string &str, const char *value)
+{
+    if (value == nullptr)
+    {
+        str.clear();
+    }
+    else
     {
-        if (value == nullptr)
```
Comment (MoKob) [inline include/extractor/extraction_way.hpp]: looks strange here, but this is actually just formated

File: include/extractor/extraction_way.hpp [hunk 4]
```diff
@@ -67,18 +66,24 @@ struct ExtractionWay
...
-    const char* GetRef() const { return ref.c_str(); }
-    void SetDestinations(const char* value) { detail::maybeSetString(destinations, value); }
-    const char* GetDestinations() const { return destinations.c_str(); }
-    void SetPronunciation(const char* value) { detail::maybeSetString(pronunciation, value); }
-    const char* GetPronunciation() const { return pronunciation.c_str(); }
-    void SetTurnLanesForward(const char* value) { detail::maybeSetString(turn_lanes_forward, value); }
-    const char* GetTurnLanesForward() const { return turn_lanes_forward.c_str(); }
-    void SetTurnLanesBackward(const char* value) { detail::maybeSetString(turn_lanes_backward, value); }
-    const char* GetTurnLanesBackward() const { return turn_lanes_backward.c_str(); }
+    void SetName(const char *value) { detail::maybeSetString(name, value); }
```
Comment (MoKob) [inline include/extractor/extraction_way.hpp]: also just formated

File: src/extractor/guidance/coordinate_extractor.cpp [hunk 5]
```diff
@@ -736,21 +816,18 @@ CoordinateExtractor::TrimCoordinatesToLength(std::vector<util::Coordinate> coord
 
     if (use_cache && length_cache.back() < desired_length && coordinates.size() >= 2)
     {
```
Comment (MoKob) [inline src/extractor/guidance/coordinate_extractor.cpp]: here, we had problems regarding the length cache. The cache was not in sync with the coordinates, resulting in potential problems when requiring both to be of the same size.

File: src/extractor/guidance/coordinate_extractor.cpp [hunk 6]
```diff
@@ -76,25 +76,48 @@ util::Coordinate CoordinateExtractor::ExtractRepresentativeCoordinate(
...
     // fallback, mostly necessary for dead ends
     if (intersection_node == to_node)
-        return TrimCoordinatesToLength(std::move(coordinates),
-                                       distance_to_skip_over_due_to_coordinate_inaccuracies)
-            .back();
+    {
+        const auto result =
+            TrimCoordinatesToLength(std::move(coordinates),
+                                    distance_to_skip_over_due_to_coordinate_inaccuracies)
+                .back();
```
Comment (daniel-j-h) [inline src/extractor/guidance/coordinate_extractor.cpp]: Hm we always generate and copy the vector around just to get its last item?

File: src/extractor/guidance/coordinate_extractor.cpp [hunk 7]
```diff
@@ -104,10 +127,13 @@ util::Coordinate CoordinateExtractor::ExtractRepresentativeCoordinate(
...
-    if (turn_edge_data.roundabout || turn_edge_data.circular )
-        return TrimCoordinatesToLength(std::move(coordinates),
-                                       distance_to_skip_over_due_to_coordinate_inaccuracies)
-            .back();
+    if (turn_edge_data.roundabout || turn_edge_data.circular)
+    {
+        coordinates = TrimCoordinatesToLength(std::move(coordinates),
+                                              distance_to_skip_over_due_to_coordinate_inaccuracies);
+        BOOST_ASSERT(is_valid_result(coordinates.back()));
+        return coordinates.back();
```
Comment (daniel-j-h) [inline src/extractor/guidance/coordinate_extractor.cpp]: Same here

File: src/extractor/guidance/coordinate_extractor.cpp [hunk 8]
```diff
@@ -255,26 +317,29 @@ util::Coordinate CoordinateExtractor::ExtractRepresentativeCoordinate(
...
         /*
          * In curves we now have to distinguish between larger curves and tiny curves modelling the
          * actual turn in the beginnig.
          *
          * We distinguish between turns that simply model the initial way of getting onto the
          * destination lanes and the ones that performa a larger turn.
          */
-        const double offset = 0.5 * considered_lanes * ASSUMED_LANE_WIDTH;
+        const double offset =
+            std::min(0.5 * considered_lanes * ASSUMED_LANE_WIDTH, 0.2 * segment_distances.back());
```
Comment (daniel-j-h) [inline src/extractor/guidance/coordinate_extractor.cpp]: Why the change?

File: src/extractor/guidance/intersection_generator.cpp [hunk 9]
```diff
@@ -44,197 +45,98 @@ Intersection IntersectionGenerator::operator()(const NodeID from_node, const Edg
...
-        }
-
-        BOOST_ASSERT(from_node == node_based_graph.GetTarget(intersection[self_u_turn].eid));
-        intersection[self_u_turn].entry_allowed = true;
+        const auto base_bearing = util::bearing::reverseBearing(intersection.begin()->bearing);
+        std::sort(intersection.begin(),
+                  intersection.end(),
+                  [base_bearing](const ConnectedRoad &lhs, const ConnectedRoad &rhs) {
+                      return util::bearing::angleBetweenBearings(base_bearing, lhs.bearing) <
+                             util::bearing::angleBetweenBearings(base_bearing, rhs.bearing);
```
Comment (daniel-j-h) [inline src/extractor/guidance/intersection_generator.cpp]: I _think_ we already have a comparator which does this (?)

If not it could make sense to provide one in a central place with a default `std::less<>` binop.

File: include/extractor/edge_based_edge.hpp [hunk 10]
```diff
@@ -60,19 +61,12 @@ inline EdgeBasedEdge::EdgeBasedEdge(const NodeID source,
...
-        }
-        return target < other.target;
-    }
-    return source < other.source;
+    const auto unidirectional = (!forward || !backward);
+    const auto other_is_unidirectional = (!other.forward || !other.backward);
+    // if all items are the same, we want to keep bidirectional edges. due to the `<` operator,
+    // preferring 0 (false) over 1 (true), we need to compare the inverse of `bidirectional`
+    return std::tie(source, target, weight, unidirectional) <
+           std::tie(other.source, other.target, other.weight, other_is_unidirectional);
```
Comment (MoKob) [inline include/extractor/edge_based_edge.hpp]: I knew that one would make you happy :)

File: include/extractor/guidance/intersection_normalizer.hpp [hunk 11]
```diff
@@ -57,12 +60,25 @@ class IntersectionNormalizer
...
     bool CanMerge(const NodeID intersection_node,
                   const Intersection &intersection,
                   std::size_t first_index,
                   std::size_t second_index) const;
 
+    // A tool called by CanMerge
+    bool InnerCanMerge(const NodeID intersection_node,
+                       const Intersection &intersection,
+                       std::size_t first_index,
+                       std::size_t second_index) const;
```
Comment (MoKob) [inline include/extractor/guidance/intersection_normalizer.hpp]: How about the explanation at CanMerge? I've added a few additional remarks here, but what was missing from the above description for you that you couldn't tell the difference? Since it's internal, there is no danger of being called from someone

File: include/extractor/node_based_edge.hpp [hunk 12]
```diff
@@ -93,9 +94,10 @@ inline NodeBasedEdge::NodeBasedEdge(NodeID source,
                                     const LaneDescriptionID lane_description_id,
                                     guidance::RoadClassification road_classification)
     : source(source), target(target), name_id(name_id), weight(weight), forward(forward),
-      backward(backward), roundabout(roundabout), circular(circular), access_restricted(access_restricted),
-      startpoint(startpoint), is_split(is_split), travel_mode(travel_mode),
-      lane_description_id(lane_description_id), road_classification(std::move(road_classification))
+      backward(backward), roundabout(roundabout), circular(circular),
+      access_restricted(access_restricted), startpoint(startpoint), is_split(is_split),
+      travel_mode(travel_mode), lane_description_id(lane_description_id),
+      road_classification(std::move(road_classification))
```
Comment (MoKob) [inline include/extractor/node_based_edge.hpp]: yes

File: src/extractor/guidance/coordinate_extractor.cpp [hunk 13]
```diff
@@ -32,7 +32,7 @@ const constexpr double LOOKAHEAD_DISTANCE_WITHOUT_LANES = 10.0;
 // smaller widths, ranging from 2.5 to 3.25 meters. As a compromise, we use
 // the 3.25 here for our angle calculations
 const constexpr double ASSUMED_LANE_WIDTH = 3.25;
-const constexpr double FAR_LOOKAHEAD_DISTANCE = 20.0;
+const constexpr double FAR_LOOKAHEAD_DISTANCE = 40.0;
```
Comment (MoKob) [inline src/extractor/guidance/coordinate_extractor.cpp]: Coordinate extraction was broken. It stopped on the first segment that was longer than 20, not the total distance. Therefore I could previously reduce it without negative effects. Turned out to be a bug.

File: src/extractor/guidance/coordinate_extractor.cpp (hunk 8 above)
Comment (MoKob) [inline src/extractor/guidance/coordinate_extractor.cpp]: Turned out to be a broken test if the segment was very small

**Comment by MoKob** [comment]
After adjustments @daniel-j-h requested:
 ```
[info] Renumbering edges: 0.249781s
[info] Generating nodes: 14.7403s
[info] Generating edges: 275.093s
```

File: include/extractor/guidance/intersection.hpp [hunk 14]
```diff
@@ -76,10 +96,24 @@ struct ConnectedRoad final : public TurnOperation
 // small helper function to print the content of a connected road
 std::string toString(const ConnectedRoad &road);
 
-struct Intersection final : public std::vector<ConnectedRoad>
+typedef std::vector<IntersectionShapeData> IntersectionShape;
+
+struct IntersectionView final : public std::vector<IntersectionViewData>
 {
-    using Base = std::vector<ConnectedRoad>;
+    typedef std::vector<IntersectionViewData> Base;
```
Comment (daniel-j-h) [inline include/extractor/guidance/intersection.hpp]: Why the typedef instead of using everywhere in this pr?

File: src/extractor/guidance/coordinate_extractor.cpp [hunk 15]
```diff
@@ -321,16 +387,89 @@ util::Coordinate CoordinateExtractor::ExtractRepresentativeCoordinate(
...
+                     [distance, &accumulated_distance, &length_cache_itr](
+                         const util::Coordinate /*coordinate*/) {
+                         const auto result = (accumulated_distance + *length_cache_itr) >= distance;
+                         if (!result)
+                         {
+                             accumulated_distance += *length_cache_itr;
+                             ++length_cache_itr;
+                         }
+                         return result;
+                     }) - 1;
```
Comment (daniel-j-h) [inline src/extractor/guidance/coordinate_extractor.cpp]: Maybe split out the lambda a bit hard to read specially with the -1

File: src/extractor/guidance/coordinate_extractor.cpp [hunk 16]
```diff
@@ -321,16 +387,89 @@ util::Coordinate CoordinateExtractor::ExtractRepresentativeCoordinate(
...
+        coordinates.end(),
+        [distance, &accumulated_distance, &last_coordinate](const util::Coordinate coordinate) {
+            const double segment_distance =
+                util::coordinate_calculation::haversineDistance(last_coordinate, coordinate);
+            const auto result = (accumulated_distance + segment_distance) >= distance;
+            if (!result)
+                accumulated_distance += segment_distance;
+
+            return result;
+        }) - 1;
```
Comment (daniel-j-h) [inline src/extractor/guidance/coordinate_extractor.cpp]: Same here - also clang-format

File: src/extractor/guidance/intersection.cpp [hunk 17]
```diff
@@ -72,6 +70,23 @@ std::string toString(const ConnectedRoad &road)
     return result;
 }
 
+IntersectionView::Base::iterator IntersectionView::findClosestTurn(double angle)
+{
+    // use the const operator to avoid code duplication
+    return begin() +
+           std::distance(cbegin(),
+                         static_cast<const IntersectionView *>(this)->findClosestTurn(angle));
```
Comment (daniel-j-h) [inline src/extractor/guidance/intersection.cpp]: Hrm I think there is `std::as_const` but this is C++17 only..
